### PR TITLE
Refactor/ilab plugin

### PR DESCRIPTION
**Pull request by achew010**

### CODE REVIEW

File: plugins/instruct-lab/configs/fast_attn.yaml [hunk 1]
```diff
@@ -0,0 +1,11 @@
+# Configurations to accelerate data packing/padding in training
+training:
+
+  # attention module configurations
+  # e.g. padding-free modifications to attention layer
+  # e.g. adding dropout to attention layer
```
Comment (fabianlim) [inline plugins/instruct-lab/configs/fast_attn.yaml:6]: please make sure comments are correct

File: plugins/instruct-lab/configs/fast_attn.yaml [hunk 2]
```diff
@@ -0,0 +1,11 @@
...
+  attention:
+
+    # this controls the confgurations for padding_free computation of attention
+    padding_free: 
+      method: "huggingface-injected"
```
Comment (fabianlim) [inline plugins/instruct-lab/configs/fast_attn.yaml:11]: as discussed, this should be called simply `"huggingface"`

File: plugins/instruct-lab/src/fms_acceleration_ilab/__init__.py [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+from .framework_plugin_padding_free import PaddingFreeAccelerationPlugin
+# from .framework_plugin_mlp_dropout import MLPDropoutAccelerationPlugin
+
+from fms_acceleration.model_patcher import ModelPatcher
+
+PATCHES = [".flash_attn"]
+PLUGIN_PREFIX = "fms_acceleration_ilab"
+
+# TODO: remove the need for the prefix
+ModelPatcher.load_patches(
```
Comment (fabianlim) [inline plugins/instruct-lab/src/fms_acceleration_ilab/__init__.py:27]: lets not use patches if it is not necessary

File: plugins/instruct-lab/src/fms_acceleration_ilab/framework_plugin_padding_free.py [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+# Third Party
+from fms_acceleration import AccelerationPlugin
+from peft import LoraConfig
+from transformers import TrainingArguments, __version__ as transformers_version, DataCollatorForSeq2Seq
+from accelerate import Accelerator
+import torch
+from types import MethodType
+from torch.utils.data import DataLoader
+
+SUPPORTED_TRANSFORMERS_VERSION = "4.43"
```
Comment (fabianlim) [inline plugins/instruct-lab/src/fms_acceleration_ilab/framework_plugin_padding_free.py:29]: call this `TRANSFORMERS_VERSION` and put comment to explain that this was the version that padding free was merged in.

File: plugins/instruct-lab/src/fms_acceleration_ilab/framework_plugin_padding_free.py [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+                ),
+            )
+        else:
+            warnings.warn(f"transformers version is equal or later \
+                than {SUPPORTED_TRANSFORMERS_VERSION}, attention forward will not be replaced.")
+
+        return model, modifiable_args
+
+    def get_callbacks_and_ready_for_train(
+        self, model: torch.nn.Module = None, accelerator: Accelerator=None
```
Comment (fabianlim) [inline plugins/instruct-lab/src/fms_acceleration_ilab/framework_plugin_padding_free.py:104]: spacing

**Review by fabianlim — CHANGES_REQUESTED** [review]

File: plugins/instruct-lab/configs/fast_attn.yaml [hunk 6]
```diff
@@ -0,0 +1,11 @@
+# Configurations to accelerate data packing/padding in training
+training:
+
```
Comment (fabianlim) [inline plugins/instruct-lab/configs/fast_attn.yaml:3]: this file should be called something else, like `instruct_lab.yaml`

**Comment by fabianlim** [comment]
Make sure go through this checklist https://github.com/foundation-model-stack/fms-acceleration/tree/main/plugins/framework#adding-new-plugins

File: plugins/instruct-lab/src/fms_acceleration_ilab/ilab_utils.py [hunk 7]
```diff
@@ -0,0 +1,40 @@
+from dataclasses import dataclass
+import warnings
```
Comment (fabianlim) [inline plugins/instruct-lab/src/fms_acceleration_ilab/ilab_utils.py:2]: this needs a header

File: plugins/instruct-lab/pyproject.toml [hunk 8]
```diff
@@ -0,0 +1,29 @@
+[build-system]
+requires = ["hatchling"]
+build-backend = "hatchling.build"
+
+[project]
+name = "fms-acceleration-ilab"
+version = '0.0.1'
+description = "FMS Acceleration using Attention Techniques"
```
Comment (fabianlim) [inline plugins/instruct-lab/pyproject.toml:8]: FMS Acceleration Plugin for Functionalities Used in Instruct Lab Training

File: plugins/instruct-lab/pyproject.toml [hunk 9]
```diff
@@ -0,0 +1,29 @@
...
+authors = [
+  {name = "Fabian Lim", email = "flim@sg.ibm.com"},
```
Comment (fabianlim) [inline plugins/instruct-lab/pyproject.toml:10]: you can add your name here also

File: plugins/instruct-lab/src/fms_acceleration_ilab/flash_attn.py [hunk 10]
```diff
@@ -0,0 +1,78 @@
...
+    position_ids = position_ids.flatten()
+    indices_q = torch.arange(position_ids.size(0), device=position_ids.device, dtype=torch.int32)
+    cu_seq_lens = torch.cat((
+        indices_q[position_ids==0],
+        torch.tensor(position_ids.size(), device=position_ids.device, dtype=torch.int32)
+        ))
+    max_length = position_ids.max()+1
+    return (query, key, value, indices_q, (cu_seq_lens, cu_seq_lens), (max_length, max_length))
+
+# we can replace with the model patcher eventually
```
Comment (fabianlim) [inline plugins/instruct-lab/src/fms_acceleration_ilab/flash_attn.py:21]: remove this comment

File: plugins/instruct-lab/src/fms_acceleration_ilab/framework_plugin_padding_free.py [hunk 11]
```diff
@@ -0,0 +1,152 @@
...
+from fms_acceleration import AccelerationPlugin
+from peft import LoraConfig
+from transformers import TrainingArguments, __version__ as transformers_version, DataCollatorForSeq2Seq
+from accelerate import Accelerator
+import torch
+from types import MethodType
+from torch.utils.data import DataLoader
+
+# This is the version where padding-free was merged into transformers
+TRANSFORMERS_VERSION = "4.43"
```
Comment (fabianlim) [inline plugins/instruct-lab/src/fms_acceleration_ilab/framework_plugin_padding_free.py:30]: Needs to change

**Review by fabianlim — APPROVED** [review]
